### ISSUE TITLE
Visual tests, Windows: Assertion in gdaldrivermanager.cpp

### ISSUE THREAD

**Issue by BergWerkGIS** [mapnik/mapnik#3093]
trying to run

```
test_visual_run.exe --jobs=%NUMBER_OF_PROCESSORS%
```

outputs:

```
ERROR 7: Assertion `FALSE' failed
in file `gdaldrivermanager.cpp', line 415
```

Does gdal require any special build flags to run the visual tests in parallel?

**Comment by springmeyer** [comment]
On unix gdal needs to be built with `./configure --with-threads` (which is also default). Not sure how this works on windows... but here is the unix configure logic: https://github.com/OSGeo/gdal/blob/3cd6fe0e3adcf7b687e92dc3b823053eda9d9c43/gdal/configure.in#L504-L595

**Comment by BergWerkGIS** [comment]
Does the `--jobs=` flag work on Linux?

First glance at gdal source suggests, that multithreading might be switched on by default: [`#define CPL_MULTIPROC_WIN32`](https://github.com/OSGeo/gdal/blob/e5f7f305dacf2d1a518c28bb2a61ebe47d2d51ea/gdal/port/cpl_multiproc.h#L44).

I've found a (rather old) reference that not all drivers support multithreading.
As [`'gdaldrivermanager.cpp', line 415`](https://github.com/OSGeo/gdal/blob/2.0/gdal/gcore/gdaldrivermanager.cpp#L403-L415) is at such place, an reason might be that one of the drivers (or just the Windows version thereof) is not multithread enabled?

@rouault any ideas about this?
We build gdal with [this script](https://github.com/mapbox/windows-builds/blob/master/scripts/build_gdal.bat).
The build target `test_visual_run.exe` is defined [here](https://github.com/mapnik/mapnik-gyp/blob/2c3b6a80631a40849ba0ab2761efad1c8c00f5a6/mapnik.gyp#L691-L718) and depends on the [mapnik target](https://github.com/mapnik/mapnik-gyp/blob/2c3b6a80631a40849ba0ab2761efad1c8c00f5a6/mapnik.gyp#L126-L232).

**Comment by rouault** [comment]
Multi-threading is enabled by default in GDAL on Windows (and Unix as well since quite a few versions). Not sure why you get this error. Looking at the code, if one exclude multithreading problems, this could happen if there were 2 driver objects sharing the same name. Perhaps a plugin lying somewhere and the same driver also compiled in GDAL main lib ?

For efficiency, I wouldn't recommend calling GDALRegisterAll() concurrently anyway.

**Comment by BergWerkGIS** [comment]
@rouault thanks for your input, `GDALAllRegister()` indeed seems to be the hot spot here.
@springmeyer  Race condition maybe? Caused by the way test runner parallelizes tests? 

`GDALAllRegister()` is called in the gdal input plugin [here](https://github.com/mapnik/mapnik/blob/7a1ab599baa1e533ece7b53ff9c1e6eafc180195/plugins/input/gdal/gdal_datasource.cpp#L61).

These are my findings so far:
* `--jobs=%NUMBER_OF_PROCESSORS%` => :fire: 
* `--jobs=%NUMBER_OF_PROCESSORS%-1` => :white_check_mark: 

Just putting `MAPNIK_LOG_ERROR(gdal) << "before GDALAllRegister()";` before the call to `GDALAllRegister()` results in this:
* `--jobs=%NUMBER_OF_PROCESSORS%` => :white_check_mark: 
* `--jobs=%NUMBER_OF_PROCESSORS%+1` => :white_check_mark: 
and works up to
* `--jobs=%NUMBER_OF_PROCESSORS%*2-2` => :white_check_mark: 
* more jobs => :fire:

**Comment by rouault** [comment]
looking more closely on how drivers are registered, if 2 threads call the registration function of the same driver at the same time, it might be racy and cause the assert you observe. So GDALAllRegister() is definitely not thead safe.

**Comment by rouault** [comment]
as more or less implied by its doc ;-) "This function should generally be called once at the beginning of the application."

**Comment by springmeyer** [comment]
Hit a case on OS X where I'm seeing a crash like this at exit. Not sure it is related to this, but dropping here for (my) later reference. (Seen using mapbox-tile-copy on a large, fully alpha compressed tiff):

```
Thread 0:: Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fff9e6d617a madvise + 10
1   libsystem_malloc.dylib        	0x00007fff93a45731 szone_free_definite_size + 5027
2   libgdal_mapnik.dylib          	0x00000001081868e6 GDALRasterBlock::~GDALRasterBlock() + 38
3   libgdal_mapnik.dylib          	0x0000000108180d45 GDALRasterBand::FlushBlock(int, int, int) + 309
4   libgdal_mapnik.dylib          	0x0000000108180e6d GDALRasterBand::FlushCache() + 269
5   libgdal_mapnik.dylib          	0x000000010814f743 GDALDataset::FlushCache() + 67
6   libgdal_mapnik.dylib          	0x000000010817587e GDALPamDataset::FlushCache() + 14
7   libgdal_mapnik.dylib          	0x0000000108094b34 GTiffDataset::Finalize() + 180
8   libgdal_mapnik.dylib          	0x0000000108095da8 GTiffDataset::CloseDependentDatasets() + 40
9   libgdal_mapnik.dylib          	0x000000010815bdbe GDALDriverManager::~GDALDriverManager() + 94
10  libgdal_mapnik.dylib          	0x000000010815c16e GDALDriverManager::~GDALDriverManager() + 14
11  libgdal_mapnik.dylib          	0x0000000108158f91 GDALDestructor() + 97
12  dyld                          	0x00007fff6807e1cc ImageLoaderMachO::doTermination(ImageLoader::LinkContext const&) + 212
13  dyld                          	0x00007fff6806d18b dyld::runAllStaticTerminators(void*) + 67
14  libsystem_c.dylib             	0x00007fff991390ff __cxa_finalize_ranges + 345
15  libsystem_c.dylib             	0x00007fff99139403 exit + 55
16  node                          	0x00000001003f5bf8 node::FatalException(v8::TryCatch&) + 302
17  node                          	0x00000001003f5a3a node::MakeDomainCallback(v8::Handle<v8::Object>, v8::Handle<v8::Function>, int, v8::Handle<v8::Value>*) + 463
18  node                          	0x00000001003f5d93 node::MakeCallback(v8::Handle<v8::Object>, v8::Handle<v8::String>, int, v8::Handle<v8::Value>*) + 73
19  node                          	0x000000010040deb8 node::StreamWrap::AfterWrite(uv_write_s*, int) + 208
20  node                          	0x000000010044602a uv__stream_io + 1967
21  node                          	0x000000010044001d uv_run + 207
22  node                          	0x00000001003f9cc8 node::Start(int, char**) + 365
23  node                          	0x0000000100001234 start + 52
```

```
Driver: GTiff/GeoTIFF
Files: hires.tif
Size is 16384, 16384
Coordinate System is:
PROJCS["WGS 84 / Pseudo-Mercator",
    GEOGCS["WGS 84",
        DATUM["WGS_1984",
            SPHEROID["WGS 84",6378137,298.257223563,
                AUTHORITY["EPSG","7030"]],
            AUTHORITY["EPSG","6326"]],
        PRIMEM["Greenwich",0],
        UNIT["degree",0.0174532925199433],
        AUTHORITY["EPSG","4326"]],
    PROJECTION["Mercator_1SP"],
    PARAMETER["central_meridian",0],
    PARAMETER["scale_factor",1],
    PARAMETER["false_easting",0],
    PARAMETER["false_northing",0],
    UNIT["metre",1,
        AUTHORITY["EPSG","9001"]],
    EXTENSION["PROJ4","+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +wktext  +no_defs"],
    AUTHORITY["EPSG","3857"]]
Origin = (7866287.454880000092089,4505504.195240000262856)
Pixel Size = (0.298582737856805,-0.298582737856805)
Metadata:
  AREA_OR_POINT=Area
Image Structure Metadata:
  COMPRESSION=LZW
  INTERLEAVE=PIXEL
Corner Coordinates:
Upper Left  ( 7866287.455, 4505504.195) ( 70d39'50.62"E, 37d28'29.49"N)
Lower Left  ( 7866287.455, 4500612.216) ( 70d39'50.62"E, 37d26'23.91"N)
Upper Right ( 7871179.434, 4505504.195) ( 70d42'28.83"E, 37d28'29.49"N)
Lower Right ( 7871179.434, 4500612.216) ( 70d42'28.83"E, 37d26'23.91"N)
Center      ( 7868733.445, 4503058.205) ( 70d41' 9.73"E, 37d27'26.71"N)
Band 1 Block=256x256 Type=Byte, ColorInterp=Red
  Mask Flags: PER_DATASET ALPHA 
Band 2 Block=256x256 Type=Byte, ColorInterp=Green
  Mask Flags: PER_DATASET ALPHA 
Band 3 Block=256x256 Type=Byte, ColorInterp=Blue
  Mask Flags: PER_DATASET ALPHA 
Band 4 Block=256x256 Type=Byte, ColorInterp=Alpha
```

**Comment by erikvullings** [comment]
I am using mapnik-tilelive on Windows 7 and 10 to render a geojson as a tile layer. This works very well, but when I try to create a second tile layer in the same mapnik-tilelive application (essentially, calling tilelive.load twice, once for each mapnik style file) I also get this error. 

Any suggestions on how to solve this?

**Comment by erikvullings** [comment]
A solution for my problem, if anyone else walks into it: I was calling tilelive.load async. Now, I use async.series, calling the tilelive.load functions one by one (only proceeding when it is ready).

**Comment by springmeyer** [comment]
This is now solved by #3395. Thanks everyone for all the testing and feedback. I confirmed by bumping up the concurrency on appveyor for the visual tests and now they all pass without crashing: https://ci.appveyor.com/project/Mapbox/mapnik/build/1.0.904#L1684. One cairo test fails but I think that is probably because cairo itself is not threadsafe, which is a different issue tracked at https://github.com/mapnik/mapnik/issues/2956
